### PR TITLE
Remove whole-frame h2 decoder

### PR DESCRIPTION
**Pull request by graebm**
RED CODE IS BEST CODE.

We use the streaming decoder now. Most changes are rewriting tests that assumed the encoder/decoder were mirrors of each other.

**REMOVE**
- `aws_h2_frame_decoder_***()` functions
- `fuzz_h2_frames.c` which just tested the old decoder.

**ALTER**
- Rewrite encoder tests (delete `test_h2_frames.c`, add `test_h2_encoder.c`). Old tests relied on whole-frame decoder.
- Alter `test_h2_headers.c` to use new streaming decoder.
- Add `aws_h2_encode()` function that accepts all frames types, instead of needing to call the write encode() function per frame type.

**FIX**
- Fix encoder bug with padding in DATA frame.
- Fix decoder bug where header-value-string overwrote header-name-string.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: source/h2_decoder.c [hunk 1]
```diff
@@ -384,6 +387,21 @@ static void s_decoder_run_state(
...
 }
 
+static void s_decoder_run_state(
+    struct aws_h2_decoder *decoder,
+    const struct decoder_state *state,
+    struct aws_byte_cursor *input) {
+
+    s_decoder_run_state_ex(decoder, state, input, false /*preserve_scratch*/);
+}
+static void s_decoder_run_state_preserve_scratch(
```
Comment (justinboswell) [inline source/h2_decoder.c]: is there a better name we could use here, maybe ```_persistent``` or ```_continue``` or ```_stream``` or just something more indicative of the actual semantics here?

File: tests/test_h2_encoder.c [hunk 2]
```diff
@@ -0,0 +1,360 @@
...
+    frame.data = aws_byte_cursor_from_c_str("hello");
+
+    /* clang-format off */
+    uint8_t expected[] = {
+        0x00, 0x00, 0x08,           /* Length (24) */
+        AWS_H2_FRAME_T_DATA,        /* Type (8) */
+        AWS_H2_FRAME_F_END_STREAM | AWS_H2_FRAME_F_PADDED, /* Flags (8) */
+        0x76, 0x54, 0x32, 0x10,     /* Reserved (1) | Stream Identifier (31) */
+        /* DATA */
+        0x02,                       /* Pad Length? (8) */
```
Comment (justinboswell) [inline tests/test_h2_encoder.c]: Is this still a ?

File: tests/test_h2_encoder.c [hunk 3]
```diff
@@ -0,0 +1,360 @@
...
+    /* Intentionally leaving header block fragment empty. Header block encoding is tested elsewhere */
+
+    /* clang-format off */
+    uint8_t expected[] = {
+        0x00, 0x00, 0x08,           /* Length (24) */
+        AWS_H2_FRAME_T_HEADERS,     /* Type (8) */
+        AWS_H2_FRAME_F_END_STREAM | AWS_H2_FRAME_F_END_HEADERS | AWS_H2_FRAME_F_PADDED | AWS_H2_FRAME_F_PRIORITY, /* Flags (8) */
+        0x76, 0x54, 0x32, 0x10,     /* Reserved (1) | Stream Identifier (31) */
+        /* HEADERS */
+        0x02,                       /* Pad Length? (8) */
```
Comment (justinboswell) [inline tests/test_h2_encoder.c]: IS this a ?

File: include/aws/http/private/h2_frames.h [hunk 4]
```diff
@@ -271,48 +265,34 @@ int aws_h2_frame_header_block_encode(
...
- */
-
-AWS_HTTP_API
-int aws_h2_frame_decoder_init(struct aws_h2_frame_decoder *decoder, struct aws_allocator *allocator);
+/* #TODO: remove each frame type's specific encode() function from API */
 AWS_HTTP_API
-void aws_h2_frame_decoder_clean_up(struct aws_h2_frame_decoder *decoder);
-AWS_HTTP_API
-int aws_h2_frame_decoder_begin(struct aws_h2_frame_decoder *decoder, struct aws_byte_cursor *data);
+int aws_h2_encode(
```
[the remark this answers is not in the document]
Comment (graebm) [reply include/aws/http/private/h2_frames.h]: 👍 will do

File: source/h2_decoder.c [hunk 5]
```diff
@@ -368,13 +368,16 @@ static enum aws_hpack_decode_status s_decode_string(
  * State functions
  **********************************************************************************************************************/
 
-static void s_decoder_run_state(
+static void s_decoder_run_state_ex(
```
[the remark this answers is not in the document]
Comment (graebm) [reply source/h2_decoder.c]: you know! "Extended"! Like Microsoft!
In my defense, this is just common functionality for the `s_decoder_run_state()` and `s_decoder_run_state_preserve_scratch()` functions. It's not called by anything else

File: source/h2_decoder.c [hunk 6]
```diff
@@ -431,19 +449,12 @@ static int s_state_fn_header(struct aws_h2_decoder *decoder, struct aws_byte_cur
...
     AWS_ASSERT(input->len >= 9);
 
-    uint32_t payload_len = 0;
-    uint8_t *length_ptr = ((uint8_t *)&payload_len);
-
     /* Read the first 3 bytes */
-    bool succ = aws_byte_cursor_read(input, length_ptr, 3);
+    uint32_t payload_len = 0;
+    bool succ = aws_byte_cursor_read_be24(input, &payload_len);
     AWS_ASSERT(succ);
```
[the remark this answers is not in the document]
Comment (graebm) [reply source/h2_decoder.c]: ¯\_(ツ)_/¯
It's a debugging assert, the programmer will go to that line and see what's wrong.

File: source/h2_decoder.c (hunk 1 above)
Comment (graebm) [reply source/h2_decoder.c]: I feel like, for this internal function, being explicit about the exact trick I'm pulling here is more helpful than referring to it with an abstract term. If we did this all the time I might invent terminology for it and use that terminology consistently

File: tests/test_h2_encoder.c (hunk 3 above)
Comment (graebm) [reply tests/test_h2_encoder.c]: I copied this from the spec, but in the spec it's part of a big ascii-art picture. I see how it's confusing here by itself. It means that it's only present when a specific FLAG is used earlier. I'll find a better way to illustrate this

Comment (graebm) [reply tests/test_h2_encoder.c]:
```
0x02,                       /* Pad Length (8)                           - F_PADDED */
0x81, 0x23, 0x45, 0x67,     /* Exclusive (1) | Stream Dependency (31)   - F_PRIORITY*/
0x09,                       /* Weight (8)                               - F_PRIORITY */
                            /* Header Block Fragment (*) */
0x00, 0x00                  /* Padding (*)                              - F_PADDED */
```

**Pull request merged**